parachains/pallets/collective-content/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::OpaqueCid) of the IPFS document of the collective charter.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::set_charter())]
+		pub fn set_charter(origin: OriginFor<T>, cid: OpaqueCid) -> DispatchResult {
+			T::CharterOrigin::ensure_origin(origin)?;
+
+			Charter::<T, I>::put(&cid);
+
+			Self::deposit_event(Event::<T, I>::NewCharterSet { cid });
+			Ok(())
+		}
+
+		/// Publish an announcement.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::OpaqueCid) of the IPFS document to announce.
+		/// - `maybe_expire`: Expiration block of the announcement.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::announce(maybe_expire.map_or(0, |_| 1)))]
+		pub fn announce(
+			origin: OriginFor<T>,
+			cid: OpaqueCid,
+			maybe_expire: Option<DispatchTime<BlockNumberFor<T>>>,
+		) -> DispatchResult {
+			T::AnnouncementOrigin::ensure_origin(origin)?;
+
+			let now = frame_system::Pallet::<T>::block_number();
+			let maybe_expire_at = maybe_expire.map(|e| e.evaluate(now));
+			ensure!(maybe_expire_at.map_or(true, |e| e > now), Error::<T, I>::InvalidExpiration);
+
+			<Announcements<T, I>>::insert(cid.clone(), maybe_expire_at);
+			<AnnouncementsCount<T, I>>::mutate(|count| *count += 1);
+
+			if let Some(expire_at) = maybe_expire_at {
+				if NextAnnouncementExpireAt::<T, I>::get().map_or(true, |n| n > expire_at) {
+					NextAnnouncementExpireAt::<T, I>::put(expire_at);
+				}
+			}
+
+			Self::deposit_event(Event::<T, I>::AnnouncementAnnounced { cid, maybe_expire_at });
+			Ok(())
+		}
+
+		/// Remove an announcement.
+		///
+		/// Parameters:
+		/// - `origin`: Must be the [Config::CharterOrigin].
+		/// - `cid`: [CID](super::OpaqueCid) of the IPFS document to remove.
+		///
+		/// Weight: `O(1)`.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::remove_announcement())]
+		pub fn remove_announcement(origin: OriginFor<T>, cid: OpaqueCid) -> DispatchResult {
+			T::AnnouncementOrigin::ensure_origin(origin)?;
+			ensure!(
+				<Announcements<T, I>>::contains_key(cid.clone()),
+				Error::<T, I>::MissingAnnouncement
+			);
+
+			<Announcements<T, I>>::remove(cid.clone());
+			<AnnouncementsCount<T, I>>::mutate(|count| *count -= 1);
+
+			Self::deposit_event(Event::<T, I>::AnnouncementRemoved { cid });
+			Ok(())
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		/// Clean up expired announcements.
+		pub fn cleanup_announcements(now: BlockNumberFor<T>) {
+			if NextAnnouncementExpireAt::<T, I>::get().map_or(true, |next| next > now) {
+				// no expired announcements expected.
+				return
+			}
+			let mut maybe_next: Option<BlockNumberFor<T>> = None;
+			let mut count = 0;
+			<Announcements<T, I>>::translate(|cid, maybe_expire_at: Option<BlockNumberFor<T>>| {
```

Comment:
Since this takes every item in the map, what if it gets too large to clean up in one block? Maybe there can be one more function to `remove_expired_announcement` that is `ensure_none` so that people can clear them.

Reply:
I introduced a default lifetime and fee free removal for any signed origin if an announcement expired.
not sure about ensure_none, since it can be spammed if none allowed.